### PR TITLE
Implemented Partition Problem, Recursive problem

### PR DESCRIPTION
**Pull request by vedas-dixit**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in their comments that points to Wikipedia or another similar explanation.
- [x] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by vedas-dixit** [comment]
Hello @appgurueu & @raklaptudirm, Can you guys please review "Tug of war-backtracking problem" that I committed recently and leave a comment if any change or modification is required :)

File: Backtracking/TugOfWar.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
+// https://www.geeksforgeeks.org/tug-of-war/
```
Comment (appgurueu) [inline Backtracking/TugOfWar.js:1]: This should be an `@see` annotation in the doc comment.

File: Backtracking/TugOfWar.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+// Given an array of n integers, we need to divide them into two subsets such that
```
Comment (appgurueu) [inline Backtracking/TugOfWar.js:3]: This also belongs in the doc comment - it describes what the function does.

File: Backtracking/TugOfWar.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+// the difference between their sums is minimized.
+
+/**
+ * @param {number[]} arr - The array of integers.
+ * @returns {Array.<number[]>} Two subsets with minimized sum difference.
+ */
+
+
+function tugOfWar(arr) {
+    const n = arr.length;
```
Comment (appgurueu) [inline Backtracking/TugOfWar.js:13]: `n` is only used twice, and one usage can be eliminated => this variable should be removed, it doesn't help readability.

File: Backtracking/TugOfWar.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    let totalSum = arr.reduce((a, b) => a + b, 0);
+    let minimumDiff = Number.MAX_VALUE;
+    let result = [];
```
Comment (appgurueu) [inline Backtracking/TugOfWar.js:16]: Why not start out with `result = [arr, []]` (put `arr` into one subset, other subset is empty), and set `minimumDiff` to `totalSum`?

File: Backtracking/TugOfWar.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * @param {number[]} arr - The array of integers.
+ * @returns {Array.<number[]>} Two subsets with minimized sum difference.
+ */
+
+
+function tugOfWar(arr) {
+    const n = arr.length;
+    let totalSum = arr.reduce((a, b) => a + b, 0);
```
Comment (appgurueu) [inline Backtracking/TugOfWar.js:14]: This should be `const`

File: Backtracking/TugOfWar.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    let minimumDiff = Number.MAX_VALUE;
+    let result = [];
+
+    function isSafe(index, subset) {
```
Comment (appgurueu) [inline Backtracking/TugOfWar.js:18]: This function is unnecessary. It just makes the code less readable; the `subset` parameter is unused.

File: Backtracking/TugOfWar.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+            if (currentDiff < minimumDiff) {
+                minimumDiff = currentDiff;
+                result = [subset, arr.filter(x => !subset.includes(x))];
+            }
+            return;
+        }
+
+        // Include the current element.
+        if (isSafe(index, subset)) {
```
Comment (appgurueu) [inline Backtracking/TugOfWar.js:36]: This check is unnecessary. `index < n` will always hold if you reach this point.

File: Backtracking/TugOfWar.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+            backtrack([...subset, arr[index]], index + 1);
+        }
+
+        // Exclude the current element.
+        if (isSafe(index, subset)) {
```
Comment (appgurueu) [inline Backtracking/TugOfWar.js:41]: Same here.

**Review by appgurueu — CHANGES_REQUESTED** [review]
I think this should be categorized as "Recursive" rather than "Backtracking". One definition of *Backtracking* is:

> Backtracking is a class of algorithms for finding solutions to some computational problems, notably constraint satisfaction problems, that incrementally builds candidates to the solutions, and *abandons a candidate as soon as it determines that the candidate cannot possibly be completed to a valid solution.*

I don't see the emphasized part here. To me, this is just a recursive algorithm which tries all possible ways to partition the set into two subsets, evaluating them and picking the best, solving an *optimization problem*.

**Comment by vedas-dixit** [comment]
I was also thinking of putting it in the "Recursive" Questions but this question is labelled as "Backtracking" so I just followed the trend.
Thanks for the review, I'll make the necessary changes :)

**Comment by vedas-dixit** [comment]
I've made some necessary changes @appgurueu 

0. Changed File Location of "TugOfWar.js" under "Recursive" from "Backtracking"
1. Removed unused variable n for clarity.
2. Initialized result with the entire array in one subset and an empty second subset, setting minimumDiff to totalSum.
3. Changed totalSum to const for immutability.
4. Deleted the redundant isSafe function and its associated checks.

**Comment by vedas-dixit** [comment]
I've made some necessary changes @appgurueu & @raklaptudirm 

Changed File Location of "TugOfWar.js" under "Recursive" from "Backtracking"
Removed unused variable n for clarity.
Initialized result with the entire array in one subset and an empty second subset, setting minimumDiff to totalSum.
Changed totalSum to const for immutability.
Deleted the redundant isSafe function and its associated checks.

**Review by appgurueu — CHANGES_REQUESTED** [review]
Ah, I see the issue now. This does not solve the "tug of war" problem; you are missing the side condition that both sets should have size n/2.

This does solve the "partition" problem (which can be reduced to the 0-1-knapsack problem) however.

I suggest either renaming this to `partition` - and removing the GfG reference - or fixing the implementation s.t. the subsets each have size n/2 (note: if n is odd, the sizes may differ by one).

**Comment by vedas-dixit** [comment]
I appreciate your feedback @appgurueu 
I've made the suggested changes to the code. It's now the "Partition Problem," and I've added error handling for invalid input. You can Check out the updated code and tests :)

**Review by appgurueu — APPROVED** [review]
(Note: Calculating the sum of nums in each recursive call is wasteful. But the implementation is correct, and really "efficiently" solving the problem isn't possible anyways (unless P = NP), so I don't mind terribly much. Ideally you should have an outer function, which takes just `nums` as a parameter, and an inner function, which takes just `index` and `target`.)

**Pull request merged**